### ISSUE TITLE
A libcurl function was given a bad argument using commands with -F, --form

### ISSUE THREAD

**Issue by jeffwitthuhn** [curl/curl#2891]
### I did this
git clone this
https://android.googlesource.com/platform/external/curl/+/android-8.1.0_r43

./buildconf
export OPENSSL_DIR=/usr/lib
export ZLIB_DIR=/usr/lib
export CFLAGS="-Wno-error"
export LDFLAGS="-L/usr/lib " #path to zlib and openssl library folder
export CC=clang
export LIBS="-lssl -lcrypto"
./configure --enable-static --prefix=/Users/jeffwitthuhn/Desktop/curltest/curlbuildmacos --with-ssl=$OPENSSL_DIR --with-zlib=$ZLIB_DIR --disable-shared 
make
make install

cd ../curlbuildmacos/bin/
./curl -F file=@/tmp example.com -vvv                                                                                                                          
* Rebuilt URL to: example.com/
*   Trying 2606:2800:220:1:248:1893:25c8:1946...
* TCP_NODELAY set
* Connected to example.com (2606:2800:220:1:248:1893:25c8:1946) port 80 (#0)
* Connection #0 to host example.com left intact
curl: (43) A libcurl function was given a bad argument

replacing -F with --data returns the html without error
./curl --data file=@/tmp example.com -vvv                                           
* Rebuilt URL to: example.com/
*   Trying 2606:2800:220:1:248:1893:25c8:1946...
* TCP_NODELAY set
* Connected to example.com (2606:2800:220:1:248:1893:25c8:1946) port 80 (#0)
> POST / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.52.0-DEV
> Accept: */*
> Content-Length: 10
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 10 out of 10 bytes
< HTTP/1.1 200 OK
< Accept-Ranges: bytes
< Cache-Control: max-age=604800
< Content-Type: text/html
< Date: Thu, 16 Aug 2018 19:12:04 GMT
< Etag: "1541025663"
< Expires: Thu, 23 Aug 2018 19:12:04 GMT
< Last-Modified: Fri, 09 Aug 2013 23:54:35 GMT
< Server: EOS (vny006/0451)
< Content-Length: 1270
<
HTML...
>
* Curl_http_done: called premature == 0
* Connection #0 to host example.com left intact                        

All other commands that use -F fail and I need to figure out how to get -F to work :(   

I'm at a bit of a loss here, I've tried several different version from https://android.googlesource.com/platform/external/curl/ but they all have the same error. 

all other curl commands without -F seem to work as expected as well.                                                                                               

### I expected the following
I expect -F to  not crash, I'm having a lot of trouble figuring this out. 

### curl/libcurl version
https://android.googlesource.com/platform/external/curl/+/android-8.1.0_r43

I need the android version to work, the same error happens when I build for android and macOS
./curl -V
curl 7.52.0-DEV (unknown) libcurl/7.52.0-DEV OpenSSL/2.0.0 zlib/1.2.11
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: IPv6 Largefile NTLM NTLM_WB SSL libz UnixSockets

### operating system
macOS 10.13.4 

adb shell getprop ro.build.version.release
7.1.1

**Comment by bagder** [comment]
Any particular reason you're building this old curl version?

The symptoms you show here really makes it look like you're having a mix of one version of curl with a different (older) version of libcurl. I've not seen anyone have this problem with a build of this or any other curl version.

**Comment by bagder** [comment]
BTW, I should add that we support curl and how to build curl with the source code *we* host and ship. I don't know how you're supposed to build what google hosts and you build like that. I think you need to ask someone into Android's builds procedure for help with that.

**Comment by jeffwitthuhn** [comment]
Same thing on newer version:  except "read function returned funny value", again --data works as expected

26|monterey:/vision/curlbuild3/bin # ./curl -V
WARNING: linker: /vision/curlbuild3/bin/curl: unsupported flags DT_FLAGS_1=0x8000001
curl 7.61.0 (aarch64-unknown-linux-android) libcurl/7.61.0 OpenSSL/1.0.2o zlib/1.2.8
Release-Date: 2018-07-11
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy
monterey:/vision/curlbuild3/bin # ./curl -F file=@/tmp example.com -vvv
WARNING: linker: /vision/curlbuild3/bin/curl: unsupported flags DT_FLAGS_1=0x8000001
Warning: setting file /tmp  failed!
* Rebuilt URL to: example.com/
*   Trying 2606:2800:220:1:248:1893:25c8:1946...
* TCP_NODELAY set
*   Trying 93.184.216.34...
* TCP_NODELAY set
* Connected to example.com (93.184.216.34) port 80 (#0)
> POST / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.61.0
> Accept: */*
> Transfer-Encoding: chunked
> Content-Type: multipart/form-data; boundary=------------------------467d04e60f57375a
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
* read function returned funny value
* Closing connection 0
curl: (26) read function returned funny value
26|monterey:/vision/curlbuild3/bin #

**Comment by jeffwitthuhn** [comment]
Hey thanks for the reply @bagder ! 

curl -V shows the same version of curl/libcurl,  any clues on how to debug? 

When I compile with master from here at github it's a little different:

android: curl -V
curl 7.61.1-DEV libcurl/7.61.1-DEV OpenSSL/1.0.2o zlib/1.2.8
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy

./curl -F file=@/tmp example.com -vvv
Warning: setting file /tmp  failed!
* Rebuilt URL to: example.com/
*   Trying 2606:2800:220:1:248:1893:25c8:1946...
* TCP_NODELAY set
* Connected to example.com (2606:2800:220:1:248:1893:25c8:1946) port 80 (#0)
> POST / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.61.1-DEV
> Accept: */*
> Transfer-Encoding: chunked
> Content-Type: multipart/form-data; boundary=------------------------4ae5c2d9cfa9c7fb
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
* read function returned funny value
* Closing connection 0
curl: (26) read function returned funny value

macOS: stuck on configure 
configure: error: /usr/lib is a bad --with-ssl prefix!

**Comment by jzakrzewski** [comment]
One funny thing though:
`./curl -F file=@/tmp example.com -vvv`
I think you're trying to upload a directory and I'm like almost sure it cannot work...

**Comment by jeffwitthuhn** [comment]
hmmm I copy/pasted 
curl -F file=@/tmp example.com
as an example usage of -F from somewhere, but I can't seem to find it now.. 

Anyway, I ended up resolving my problems by getting a fresh openssl on macos from brew. This resolved the example command on macos build, so something was broken with my openssl lib I guess that caused a crash in curl :'( 

However this didn't resolve android sadly, so I ended up compiling boringssl for android and linking with that instead.  
https://android.googlesource.com/platform/external/boringssl
https://ec.haxx.se/building-boringssl.html

Maybe openssl doesn't play nice with android?

Something still isn't built correctly since I'm missing everything to do with curl_mime, and some other commands fail that shouldn't, but I can now use the curl commands I wanted at least.
